### ISSUE TITLE
Missing Example: How To Mint NFTs with Two Different Sets of Metadata

### ISSUE THREAD

**Issue by johnnyshankman** [oxheadalpha/smart-contracts#77]
It's unclear how exactly you mint two different singular NFTs on the same contract with different metadata, which is usually the case. 

This case only shows the odd case where you want to mint two NFTs with the same metadata:
https://github.com/oxheadalpha/smart-contracts/blob/master/nft_assets/tezos_fa2_nft_tests/test_fa2.py#L132-L153

Is this how?
```
        self.fa2.mint_tokens(
            {
                "metadata": token_metadata1,
                "token_def": {
                    "from_": 0,
                    "to_": 1,
                },
                "owners": [owner1_address],
            }
        self.fa2.mint_tokens(
            {
                "metadata": token_metadata2,
                "token_def": {
                    "from_": 1,
                    "to_": 2,
                },
                "owners": [owner1_address],
            }
```

**Comment by johnnyshankman** [comment]
Although I can't manage to write a working lambda, it seems from this
```
(** Finds a definition of the token type (token_id range) associated with the provided token id *)
let find_token_def (tid, token_defs : token_id * (token_def set)) : token_def =
  let tdef = Set.fold (fun (res, d : (token_def option) * token_def) ->
    match res with
    | Some r -> res
    | None ->
      if tid >= d.from_ && tid < d.to_
      then  Some d
      else (None : token_def option)
  ) token_defs (None : token_def option)
  in
  match tdef with
  | None -> (failwith fa2_token_undefined : token_def)
  | Some d -> d
```

That it seems to match on `tid` being greater than or equal to the `from`, but less than the `to` (exclusive).

So if I have token 3 and there are two pairs in storage
```
Pair(2, 3) -> url1
Pair(3, 4) -> url2
```

It will always pick Pair(3,4), as
```
3 >= 3 is true
3 < 4  is true
```

**Comment by johnnyshankman** [comment]
I think we should put this more clearly somewhere as an example for future devs.

**Comment by emishur** [comment]
> The token_metadata endpoint is also a little arcane, as it asks for a lambda as the first parameter and I'm not sure what sort of function it expects.

This is a special way to implement contract view entry points. Tezos will have Solidity-like view entry points with the next protocol update. At the moment, view entry points use continuation passing style pattern when the parameter contains a callback lambda. 

If the contract A wants to get some value from the contract B, it should call contract's A view entry point passing parameters (like token_id) and a callback. Contract B then will call a provided callback entry point of the contract A, passing the requested value.

In general, view entry points cannot be called off-chain (only by other contracts). But Taquito library has special API which lets to inspect view entry points off-chain: https://tezostaquito.io/docs/lambda_view/

**Comment by emishur** [comment]
> It's unclear how you mint two different singular NFTs on the same contract with different metadata ...
> Everything is catered toward a range of NFT IDs with the same metadata.

Yes, this is a special NFT contract which represents "semi-fungible" NFTs.
You might want to look at NFT collectibles example: https://github.com/oxheadalpha/smart-contracts/tree/master/collectibles
https://github.com/oxheadalpha/smart-contracts/blob/master/collectibles/ligo/src/fa2_single_asset.mligo contract can be originated with a fixed NFT collection where each NFT has different metadata.

You may also have a look at https://github.com/tqtezos/minter-sdk which has a variety of fungible and non-fungible token contracts implementation or take the NFT contract from https://github.com/oxheadalpha/nft-tutorial
